src/parquet/bloom_filter.cc

```diff
@@ -52,54 +47,50 @@ BloomFilter::BloomFilter(uint32_t num_bytes)
-      this->bloom_filter_algorithm_.reset(new BlockBasedAlgorithm(num_bytes));
-      break;
-    default:
-      throw parquet::ParquetException("Unsupported Bloom filter algorithm");
-  }
-}
+  num_bytes_ = num_bytes;
+  PARQUET_THROW_NOT_OK(::arrow::AllocateBuffer(pool_, num_bytes_, &data_));
+  memset(data_->mutable_data(), 0, num_bytes_);
 
-BloomFilter::BloomFilter(const uint8_t* bitset, uint32_t num_bytes)
-    : hash_strategy_(HashStrategy::MURMUR3_X64_128), algorithm_(Algorithm::BLOCK) {
-  switch (hash_strategy_) {
-    case HashStrategy::MURMUR3_X64_128:
-      this->hasher_.reset(new MurmurHash3());
-      break;
-    default:
-      throw parquet::ParquetException("Unsupported hash strategy");
-  }
+  this->hasher_.reset(new MurmurHash3());
+}
 
+BlockSplitBloomFilter::BlockSplitBloomFilter(const uint8_t* bitset, uint32_t num_bytes)
+    : pool_(::arrow::default_memory_pool()),
+      hash_strategy_(HashStrategy::MURMUR3_X64_128),
+      algorithm_(Algorithm::BLOCK) {
   if (!bitset) {
-    throw parquet::ParquetException("bitset pointer is NULL");
+    throw ParquetException("Given bitset is NULL");
   }
 
   if (num_bytes < MINIMUM_BLOOM_FILTER_BYTES || num_bytes > MAXIMUM_BLOOM_FILTER_BYTES ||
       (num_bytes & (num_bytes - 1)) != 0) {
-    throw parquet::ParquetException("Given length of bitset is illegal");
+    throw ParquetException("Given length of bitset is illegal");
   }
 
-  switch (algorithm_) {
-    case Algorithm::BLOCK:
-      this->bloom_filter_algorithm_.reset(new BlockBasedAlgorithm(bitset, num_bytes));
-      break;
-    default:
-      throw parquet::ParquetException("Unsupported Bloom filter algorithm");
-  }
+  num_bytes_ = num_bytes;
+  PARQUET_THROW_NOT_OK(::arrow::AllocateBuffer(pool_, num_bytes_, &data_));
+  memcpy(data_->mutable_data(), bitset, num_bytes_);
+
+  this->hasher_.reset(new MurmurHash3());
 }
 
-BloomFilter* BloomFilter::Deserialize(InputStream* input) {
+BloomFilter* BlockSplitBloomFilter::Deserialize(InputStream* input) {
   int64_t bytes_available;
+  const uint8_t* read_buffer = NULL;
 
   uint32_t len;
-  memcpy(&len, input->Read(sizeof(uint32_t), &bytes_available), sizeof(uint32_t));
-  if (static_cast<uint32_t>(bytes_available) != sizeof(uint32_t)) {
+  read_buffer = input->Read(sizeof(uint32_t), &bytes_available);
+  if (static_cast<uint32_t>(bytes_available) != sizeof(uint32_t) || !read_buffer) {
     throw ParquetException("Failed to deserialize from input stream");
   }
+  memcpy(&len, read_buffer, sizeof(uint32_t));
 
   uint32_t hash;
-  memcpy(&hash, input->Read(sizeof(uint32_t), &bytes_available), sizeof(uint32_t));
+  read_buffer = input->Read(sizeof(uint32_t), &bytes_available);
+  if (static_cast<uint32_t>(bytes_available) != sizeof(uint32_t) || !read_buffer) {
+    throw ParquetException("Failed to deserialize from input stream");
+  }
+  memcpy(&hash, read_buffer, sizeof(uint32_t));
   if (static_cast<uint32_t>(bytes_available) != sizeof(uint32_t)) {
```

Comment:
This is duplicated with the code on line 90. Same for line 107 and 103, respectively.